feat(cloudfront): warn users that minimumProtocolVersion defaults to TLSv1 without a custom certificate

AkitoAndo opened the pull request:
Issue # (if applicable)
Closes https://github.com/aws/aws-cdk/issues/35404

Reason for this change
Currently, AWS CDK allows setting minimumProtocolVersion on a CloudFront Distribution without requiring a custom SSL/TLS certificate. However, CloudFront silently ignores this setting when using its default certificate, defaulting to TLSv1 regardless of the configured value. This creates a security vulnerability where developers believe their distribution enforces TLS 1.2+ when it actually accepts TLS 1.0 and 1.1.

Description of changes
Added validation in the Distribution constructor to throw an error when minimumProtocolVersion is specified without a custom certificate
Added validation for sslSupportMethod as well, as it has the same dependency on custom certificates
The validation ensures that developers are explicitly aware that these security settings require a custom certificate to take effect
Added comprehensive unit tests to verify the validation logic
The validation logic checks if either minimumProtocolVersion or sslSupportMethod are set without a certificate, and throws a descriptive error message guiding users to provide a custom certificate.

Alternative considered: Adding a warning instead of an error for backward compatibility. However, given the security implications of this misconfiguration, a breaking change with clear error messaging was deemed more appropriate to prevent security vulnerabilities.

Describe any new or updated permissions being added
No new or updated IAM permissions are required for this change.

Description of how you validated changes
Added unit tests to verify that an error is thrown when minimumProtocolVersion is set without a certificate
Added unit tests to verify that an error is thrown when sslSupportMethod is set without a certificate
Added unit tests to confirm that no error is thrown when these properties are set with a valid certificate
Verified that existing tests pass with the new validation
Manually tested the changes by attempting to deploy a Distribution with the problematic configuration and confirming the error is raised during synthesis
Checklist
 My code adheres to the [CONTRIBUTING GUIDE](https:/ /github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and
[DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license

aws-cdk-automation reviewed (dismissed):
(This review is outdated)

AkitoAndo commented on packages/aws-cdk-lib/aws-cloudfront/lib/distribution.ts:
```diff
@@ -369,6 +369,16 @@ export class Distribution extends Resource implements IDistribution {
+    // Validate that minimumProtocolVersion and sslSupportMethod are only specified with a certificate
+    if (!props.certificate) {
+      if (props.minimumProtocolVersion !== undefined) {
+        throw new ValidationError('minimumProtocolVersion can only be specified when using a custom certificate. Use the \'certificate\' property to provide a certificate.', this);
```
Thank you for the feedback! I've updated the implementation to use `Annotations.addWarningV2()` instead of throwing a ValidationError.

AkitoAndo commented:
@dgandhi62 Thanks for the review! I've updated the PR based on your feedback:

- Changed ValidationError to warnings for backward compatibility
- Added explicit default behavior in warning messages (TLSv1, SNI)
- Added assertions to the integration test
- Updated all unit tests accordingly

 
 Ready for re-review when you have time. Thank you 🙇

aws-cdk-automation reviewed (dismissed):
(This review is outdated)

aws-cdk-automation commented:
<!--STALE PR-->
This PR has been in the CHANGES REQUESTED state for 3 weeks, and looks abandoned. Note that PRs with failing linting check or builds are not reviewed, please ensure your build is passing 

To prevent automatic closure:
- Resume work on the PR
- OR request an exemption by adding a comment containing 'Exemption Request' with justification e.x "Exemption Request: <justification>"
- OR request clarification by adding a comment containing 'Clarification Request' with a question e.x "Clarification Request: <question>"

This PR will automatically close in 14 days if no action is taken.
